lib/axiom/fs/stream/memory_stream_buffer.js

```diff
@@ -24,114 +24,33 @@ var ReadableStream;
 /** @typedef WritableStream$$module$axiom$fs$stream$writable_stream */
 var WritableStream;
 
-/** @typedef function():void */
-var EventCallback;
-
-/**
- * Event value + associated callback
- * @constructor
- * @param {!*} value
- * @param {!EventCallback} callback
- */
-var EventWithCallback = function(value, callback) {
-  this.value = value;
-  this.callback = callback;
-};
-
 /**
- * @constructor
+ * @constructor @extends {StreamBuffer}
  */
 export var MemoryStreamBuffer = function() {
-  /** @private @type {boolean} */
-  this.paused_ = true;
-  /** @const @private @type {Queue} */
-  this.events_ = new Queue();
-  /** @type {!AxiomEvent} */
-  this.onData = new AxiomEvent();
-  /** @type {!AxiomEvent} */
-  this.onReadable = new AxiomEvent();
+  StreamBuffer.call(this);
+
   /** @const @type {!ReadableStream} */
   this.readableStream = new ReadableMemoryStream(this);
   /** @const @type {!WritableStream} */
   this.writableStream = new WritableMemoryStream(this);
+  
+  this.onClose = AxiomEvent.all([
+      this.readableStream.onClose, 
+      this.writableStream.onClose
+  ]);
 };
 
 export default MemoryStreamBuffer;
 
-/**
- * @return {void}
- */
-MemoryStreamBuffer.prototype.pause = function() {
-  this.paused_ = true;
-};
+MemoryStreamBuffer.prototype = Object.create(StreamBuffer.prototype);
 
 /**
+ * @override
  * @return {void}
  */
-MemoryStreamBuffer.prototype.resume = function() {
-  this.paused_ = false;
-  this.flushEvents_();
-};
-
-/**
- * Consume one event from the stream. Return undefined if the
- * stream is empty.
- *
- * @return {*}
- */
-MemoryStreamBuffer.prototype.read = function() {
-  if (!this.paused_) {
-    throw new AxiomError.Runtime('Cannot read: stream must be paused');
-  }
-  return this.read_();
-};
-
-/**
- * Write data into the stream (for implementers of the stream only).
- * Invoke the "onData" callback for all pending events if the callback is set.
- *
- * @param {!*} value
- * @param {EventCallback=} opt_callback
- * @return {void}
- */
-MemoryStreamBuffer.prototype.write = function(value, opt_callback) {
-  var item = value;
-  if (opt_callback) {
-    item = new EventWithCallback(value, opt_callback);
-  }
-  this.events_.enqueue(item);
-  if (this.paused_) {
-    this.onReadable.fire();
-  } else {
-    this.flushEvents_();
-  }
-};
-
-/**
- * @return {void}
- */
-MemoryStreamBuffer.prototype.flushEvents_ = function() {
-  while (!this.paused_) {
-    var item = this.read_();
-    if (!item)
-      break;
-
-    this.onData.fire(item);
-  }
-};
-
-/**
- * Consume one event from the stream. Return undefined if the
- * stream is empty.
- *
- * @return {*}
- */
-MemoryStreamBuffer.prototype.read_ = function() {
-  var item = this.events_.dequeue();
-  if (item instanceof EventWithCallback) {
-    item.callback();
-    return item.value;
-  } else {
-    return item;
-  }
+MemoryStreamBuffer.prototype.close = function() {
```

Comment:
Oh I see... So, see my comments above: this should be named `end`, and should flush all writes before firing `onEnd` and `onFinish`.